### PR TITLE
#1103 

### PR DESCRIPTION
**Pull request by navpil**
Try to fix the issue #1103 "1.2.0-Beta1 Automapping Feature: cycle tree let the compiler crash with StackOverflowError."

Initially when there was a non-recursive implementation there was a HashSet which excluded already created Forged Mappings. With the recursive implementation I see no way except using some global state or passing the set of created forged methods as a parameter.
I decided to use the first solution. This fix is added only for the usual property mappings, I did not check how would the recursive iterable mappings behave.

Is this a fine approach?

### CODE REVIEW

**Comment by gunnarmorling** [comment]
ok to test

**Comment by agudian** [comment]
I vaguely remember that we check the available methods if they already match the soure/target type before creating a forged method, but in the near past added the mapping options to the hashCode/equals for forged methods. Is that maybe a spot where we can fix this as well? @filiphr?

**Comment by agudian** [comment]
Btw, @navpil, huge thanks for keeping in touch with your feature 👍

**Comment by filiphr** [comment]
@navpil You are right we changed a bit how the `ForgedMethods` are created. I also started working on something yesterday evening. It is a similar approach to yours. Have a look at my https://github.com/filiphr/mapstruct/commit/ebc1920eb87dc4b6c3e0bbd2d0ded9a2cede11dc. I am using a map, because the name of a `ForgedMethod` is not used in the `equals`/`hashCode` methods and we always generate a safe name for the methods. I think that with a set we could theoretically create assignments that don't match the created methods.

My fix spot is also at another place, let me know what you think.

**Comment by navpil** [comment]
@filiphr Actually the name of the ForgedMethod is used in the hashCode and equals. Anyway - map keys and sets work the same (HashSet is actually a HashMap with no values) so our approaches are quite the same. The spot is almost the same too - only a method call apart. 
The only real difference I see is that you remove the ForgedMethod from the UnderCreation collection and I don't. I don't know which one is better. What will happen if in some other place, not in the recursion path, we decide to generate another forged method with the same parameters? Should we know that such method was already created, can it cause problems if we don't create another forged method, or is it irrelevant?

**Comment by filiphr** [comment]
@navpil You are right the `ForgedMethod` does use the name in the `equals`/`hashCode` methods. However, I think that it shouldn't. We removed the name from those methods for some other `Mapping` methods. I used a `Map` only because I thought that the `name` was not used in the `equals`/`hashCode` methods.

I did it in that spot because there are other places apart from the `PropertyMapping` that create forged mappings.


>What will happen if in some other place, not in the recursion path, we decide to generate another forged method with the same parameters? 

If we generate a `ForgedMethod` with the same parameters then we check the created `MappingMethod` in `createForgedAssignment()` and if one already exists (`name` is not taken into consideration) then we just change the `ForgedMethod` that we are going to use in the `assignment`.

>Should we know that such method was already created, can it cause problems if we don't create another forged method, or is it irrelevant?

I don't think that it matters whether we create a different `ForgedMethod` or not. 

Btw. I decided to remove it so I can check in the `MapperCreationProcessor` if all `ForgedMethod`(s) finished with creation successfully (I am not sure if this is needed, I just thought as a safeguard that will notify us if something is not OK).

**Comment by sjaakd** [comment]
> @navpil You are right the ForgedMethod does use the name in the equals/hashCode methods. However, I think that it shouldn't. We removed the name from those methods for some other Mapping methods. I used a Map only because I thought that the name was not used in the equals/hashCode methods.

The reason here: we want MapStruct to be able to reuse prior generated methods. That is, their signature is the same and their content (property mappings, incl. their order) is the same. The name not matter. Note that generating the same method with different names can lead to problems (conflicting methods in the selection process).

**Comment by navpil** [comment]
@filiphr So it looks like your solution is closer to what we need than mine. Putting the check into the AbstractBaseBuilder and clearing the UnderCreation collection do make sense. 
Next the 'name' should be removed from the ForgedMethod equals/hashCode and I'd really suggest to use the Set instead of Map (Map is kind of confusing, since we don't use values). 
Should I adjust my PR to reflect your changes, or we can just replace it with yours?
What about other cases? I will look into the case where user has a list of users as a property. Other suggestions?

**Comment by navpil** [comment]
I checked that with with @filiphr solution there will be no problem with mapping tree like objects (when user has a list of users himself).

**Comment by filiphr** [comment]
@navpil feel free to adjust the PR with the changes. We don't have to replace it with mine, use what you need from there.

As for the `Set` vs `Map`. If we remove the `name` from the `equals` and `hashCode` checks, then we must use `Map`. In my "solution" the value of the `underCreation` map is used to create an assignment. If we only use `Set` then you can theoretically create an `assignment` that won't compile.

For example:
* You have `ForgedMethod` with name `carToCarDto`, you put this method into the `Map` with itself as key and value. 
* In the next step you have a `ForgedMethod` with name `carToCarDto1`. 

Both methods are equal and have the same hashCode. If you use a `Set` then you will use the name of the second method `carToCarDto1`, but you are going to create a mapping method for `carToCarDto` (the first method is used to create the mapping). If on the other side we use a `Map`, then when you use the second method to get a value you will get back the first `ForgedMethod` and you are going to create an assignment with the name `carToCarDto`.

WDYT?

**Comment by navpil** [comment]
@filiphr Finally I got it. Sorry, I did not notice you actually use the map values, so that's why I was confused about map vs set difference. As for PR - I will update it to reflect your changes.

**Comment by navpil** [comment]
I had to change how the `equals/hashCode` is being generated in the `Parameter` class. It took only name into account which caused `ForgedMethod` equality be wrong. So now it takes name and type into account. I was not sure whether name should have been removed or not, so I left it there.
This change broke one test, in particular `NestedTargetPropertiesTest`. But the reason was not that the generated mapper was wrong, but that the method names were not what as expected. I guess the reason was that the `hashCode` change changed the way objects are stored in the set buckets and were iterated through in the other order than before. So created mapper was different from the reference mapper `fixtures\org\mapstruct\ap\test\nestedtargetproperties\ChartEntryToArtistImpl.java` in particular - methods `chartEntryTopSong1` and `chartEntryToSong2` were swapped.
You can see I commented out the Rule in the test and the test itself passed.
What should be done with that?

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java [hunk 1]
```diff
@@ -84,6 +84,7 @@
     GENERAL_UNSUPPORTED_DATE_FORMAT_CHECK( "No dateFormat check is supported for types %s, %s" ),
     GENERAL_VALID_DATE( "Given date format \"%s\" is valid.", Diagnostic.Kind.NOTE ),
     GENERAL_INVALID_DATE( "Given date format \"%s\" is invalid. Message: \"%s\"." ),
+    GENERAL_NOT_ALL_FORGED_CREATED( "Internal Error in e creation of Forged Methods, it was expected all Forged Methods to finished with creation, but %s did not" ),
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Message.java]: `Error in e creation` --> I guess the " e " can be removed...

File: processor/src/test/java/org/mapstruct/ap/test/nestedtargetproperties/NestedTargetPropertiesTest.java [hunk 2]
```diff
@@ -51,11 +50,11 @@
...
-    @Rule
-    public GeneratedSource generatedSource = new GeneratedSource().addComparisonToFixtureFor(
-        ChartEntryToArtist.class,
-        ChartEntryToArtistUpdate.class
-    );
+//    @Rule
+//    public GeneratedSource generatedSource = new GeneratedSource().addComparisonToFixtureFor(
+//        ChartEntryToArtist.class,
+//        ChartEntryToArtistUpdate.class
+//    );
```
Comment (agudian) [inline processor/src/test/java/org/mapstruct/ap/test/nestedtargetproperties/NestedTargetPropertiesTest.java]: Hmm, that's not good. Can you find out where we're iterating over a map or set that doesn't contain any ordering? Typically, we use `LinkedHashMap` or somehting similar that preserves the `put` order when iterating over it.

File: processor/src/main/java/org/mapstruct/ap/internal/model/AbstractBaseBuilder.java [hunk 3]
```diff
@@ -56,12 +56,23 @@ public B method(Method sourceMethod) {
...
     Assignment createForgedBeanAssignment(SourceRHS sourceRHS, ForgedMethod forgedMethod) {
+
+        if ( ctx.getForgedMethodsUnderCreation().containsKey( forgedMethod ) ) {
+            return createAssignment( sourceRHS, ctx.getForgedMethodsUnderCreation().get( forgedMethod ) );
+        }
+        else {
+            ctx.getForgedMethodsUnderCreation().put( forgedMethod, forgedMethod );
+        }
+
+
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/model/AbstractBaseBuilder.java]: One empyt line is sufficient... 😁

File: processor/src/test/java/org/mapstruct/ap/test/nestedtargetproperties/NestedTargetPropertiesTest.java (hunk 2 above)
Comment (agudian) [inline processor/src/test/java/org/mapstruct/ap/test/nestedtargetproperties/NestedTargetPropertiesTest.java]: @navpil now that I've read your comment above a bit more careful: if you can make sure that the order remains "stable" in a sense that it's not depending on any HashMap/HashSet iteration, then it's totally okay to adapt the reference impl file.
But if this is a result of HashMap/HashSet iteration (which can vary based on the JDK version, among other things), then we should somehow fix it.

Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/nestedtargetproperties/NestedTargetPropertiesTest.java]: @navpil the problem are the `HashMap`(s) used in the `NestedTargetPropertyMappingHolder`. If you change them to `LinkedHashMap`(s) then all is good.

File: processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java [hunk 4]
```diff
@@ -18,17 +18,11 @@
  */
 package org.mapstruct.ap.internal.model;
 
-import static org.mapstruct.ap.internal.model.assignment.Assignment.AssignmentType.DIRECT;
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/PropertyMapping.java]: Apart from the imports and one empty line there are no changes to this file. Can you please revert all changes to this one?

File: processor/src/main/java/org/mapstruct/ap/internal/model/common/Parameter.java [hunk 5]
```diff
@@ -95,25 +95,28 @@ public boolean isMappingContext() {
...
+        if ( name != null ? !name.equals( parameter.name ) : parameter.name != null ) {
             return false;
         }
-        return true;
+        return type != null ? type.equals( parameter.type ) : parameter.type == null;
+
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/common/Parameter.java]: Why change the order of the `equals` / `hashCode`? Why not just add the needed lines to the methods?

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/FormattingParameters.java [hunk 6]
```diff
@@ -40,4 +40,28 @@ public String getNumber() {
         return number;
     }
 
+    @Override
+    public boolean equals(Object o) {
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/source/FormattingParameters.java]: Why are these needed for this class? Where are they used?

File: processor/src/main/java/org/mapstruct/ap/internal/util/Message.java [hunk 7]
```diff
@@ -100,8 +101,8 @@
     RETRIEVAL_TYPE_VAR_RESULT( "Can't generate mapping method for a generic type variable target." ),
     RETRIEVAL_WILDCARD_SUPER_BOUND_SOURCE( "Can't generate mapping method for a wildcard super bound source." ),
     RETRIEVAL_WILDCARD_EXTENDS_BOUND_RESULT( "Can't generate mapping method for a wildcard extends bound result." ),
-    RETRIEVAL_CONTEXT_PARAMS_WITH_SAME_TYPE( "The types of @Context parameters must be unique." ),
```
Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/util/Message.java]: Why are there line changes for this one and `INHERITCONFIGURATION_CYCLE`?

File: processor/src/test/java/org/mapstruct/ap/test/nestedbeans/RecursionTest.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    @IssueKey("1103")
+    public void testRecursiveAutoMap() {
+        RecursionMapper.Root root = RecursionMapper.INSTANCE.mapRoot(
+            new RecursionMapper.RootDto(
+                new RecursionMapper.ChildDto( "Sub Root", new RecursionMapper.ChildDto(
+                    "Sub child", null
+                ) )
+            ) );
+
+        Assert.assertEquals(
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/nestedbeans/RecursionTest.java]: Can you use object accessors instead of the equals method for the checks? You can have a look at the latest `NestedSimpleBeansMappingTest`.

Then you won't need the `equals` and `hashCode` methods in the objects and the class will be smaller :)

File: processor/src/test/java/org/mapstruct/ap/test/nestedbeans/recursive/RecursionMapper.java [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.mapstruct.ap.test.nestedbeans.recursive;
+
+import org.mapstruct.Mapper;
+import org.mapstruct.ReportingPolicy;
+import org.mapstruct.factory.Mappers;
+
+@Mapper(unmappedTargetPolicy = ReportingPolicy.IGNORE)
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/nestedbeans/recursive/RecursionMapper.java]: I think that the `unmappedTargetPolicy` is not needed.

File: processor/src/test/java/org/mapstruct/ap/test/nestedbeans/recursive/TreeRecursionMapper.java [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+ */
+package org.mapstruct.ap.test.nestedbeans.recursive;
+
+import java.util.List;
+
+import org.mapstruct.Mapper;
+import org.mapstruct.ReportingPolicy;
+import org.mapstruct.factory.Mappers;
+
+@Mapper(unmappedTargetPolicy = ReportingPolicy.IGNORE)
```
Comment (filiphr) [inline processor/src/test/java/org/mapstruct/ap/test/nestedbeans/recursive/TreeRecursionMapper.java]: Same here. I am don't think the target policy is needed.

File: processor/src/main/java/org/mapstruct/ap/internal/model/source/FormattingParameters.java (hunk 6 above)
Comment (navpil) [inline processor/src/main/java/org/mapstruct/ap/internal/model/source/FormattingParameters.java]: True to say I blindly copypasted this from the changeset you linked earlier. I will remove it and check that all tests still pass.

Comment (filiphr) [inline processor/src/main/java/org/mapstruct/ap/internal/model/source/FormattingParameters.java]: No problem. I didn't realize I have committed that as well, I was in a hurry to commit and push then and might have pushed something not needed :)

**Comment by filiphr** [comment]
@navpil will you have time to finish this PR? We want to release Beta2 and this is the last issue that is planned for it. If you don't have the time, let us know, so we can finish it.

**Comment by agudian** [comment]
@navpil do you have some time rebase and work in the remaining comments? We're planning to do the Beta2 release tonight, but I'd really like to see this fix included.
If you already know that you can't do it in time, one of us can take over, but it's usually nicer for everyone if we don't steal your credit 😉. Just let us know... 👍

**Comment by navpil** [comment]
I will do this today

File: processor/src/main/java/org/mapstruct/ap/internal/model/common/Parameter.java (hunk 5 above)
Comment (navpil) [inline processor/src/main/java/org/mapstruct/ap/internal/model/common/Parameter.java]: I usually delete equals/hashCode and generate them from the IntelliJ template instead of adding needed lines, I did not notice the methods were swapped.

File: processor/src/test/java/org/mapstruct/ap/test/nestedbeans/RecursionTest.java (hunk 8 above)
Comment (navpil) [inline processor/src/test/java/org/mapstruct/ap/test/nestedbeans/RecursionTest.java]: I will do how you ask, though as I think of it - the best solution here would be map dto to entity and back (we assume they have the same number of properties) and then compare them by the equals method. Something like
`RootDto rootDto = createTestRootDto();
assertEquals(rootDto, mapper.mapEntityToDto( mapper.mapDtoToEntity( rootDto ) );`
This way there is no need to write own equality methods (possibly error prone) and no need to create an additional object of different type, but same parameters.

**Comment by agudian** [comment]
Awesome, @navpil, thanks a lot! Release is coming up now... 👍
